x-pack/plugins/upgrade_assistant/server/lib/reindexing/index_settings.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { flow, omit } from 'lodash';
+import { ReindexWarning } from '../../../common/types';
+import { FlatSettings, MappingProperties, TypeMapping } from './types';
+
+/**
+ * Validates, and updates deprecated settings and mappings to be applied to the
+ * new updated index.
+ */
+export const transformFlatSettings = (flatSettings: FlatSettings) => {
+  const settings = transformSettings(flatSettings.settings);
+  const mappings = transformMappings(flatSettings.mappings);
+
+  return { settings, mappings };
+};
+
+/**
+ * Returns an array of warnings that should be displayed to user before reindexing begins.
+ * @param flatSettings
+ */
+export const getReindexWarnings = (flatSettings: FlatSettings): ReindexWarning[] => {
+  const mapping = getSingleMappingType(flatSettings.mappings);
+
+  const warnings = [
+    [ReindexWarning.allField, Boolean(mapping && mapping._all && mapping._all.enabled)],
+    [
+      ReindexWarning.booleanFields,
+      Boolean(mapping && mapping.properties && findBooleanFields(mapping.properties).length > 0),
+    ],
+  ] as Array<[ReindexWarning, boolean]>;
+
+  return warnings.filter(([_, applies]) => applies).map(([warning, _]) => warning);
+};
+
+/**
+ * Returns an array of field paths for all boolean fields, where each field path is an array of strings.
+ * Example:
+ *    For the mapping type:
+ *    ```
+ *      {
+ *        "field1": { "type": "boolean" },
+ *        "nested": {
+ *          "field2": { "type": "boolean" }
+ *        }
+ *      }
+ *    ```
+ *    The fieldPaths would be: `[['field1'], ['nested', 'field2']]`
+ * @param properties
+ */
+export const findBooleanFields = (properties: MappingProperties): string[][] =>
+  Object.keys(properties).reduce(
+    (res, propertyName) => {
+      if (properties[propertyName].type === 'boolean') {
+        // If this field is a boolean, add it
+        res.push([propertyName]);
+      } else if (properties[propertyName].properties) {
+        // If this is a nested object/array get the nested fields and prepend the field path with the current field.
+        const nested = findBooleanFields(properties[propertyName].properties!);
+        res = [...res, ...nested.map(n => [propertyName, ...n])];
+      }
+
+      return res;
+    },
+    [] as string[][]
+  );
+
+const removeUnsettableSettings = (settings: FlatSettings['settings']) =>
+  omit(settings, [
+    'index.uuid',
+    'index.blocks.write',
+    'index.creation_date',
+    'index.legacy',
+    'index.mapping.single_type',
+    'index.provided_name',
+    'index.routing.allocation.initial_recovery._id',
+    'index.version.created',
+    'index.version.upgraded',
+  ]);
```

Comment:
Can you also add `index.mapping.single_type`

The upgrade assistant set this in 5.6 and it's not a setting available in 6.x+

Reply:
Actually just added that a couple hours ago.